File: src/main/java/com/baidu/algorithm/dfs/_207_Course_Schedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """

        if numCourses < 2:
            return True

        graph = [[] for _ in xrange(numCourses)]
        for x, y in prerequisites:
            graph[y].append(x)

        visited = [False] * numCourses
        for i in xrange(numCourses):
            if not self.dfs(graph, visited, i):
                return False

        return True

    def dfs(self, graph, visited, i):
        if visited[i]:
            return False

        visited[i] = True
        for x in graph[i]:
            if not self.dfs(graph, visited, x):
                return False

        visited[i] = False

        return True


if __name__ == '__main__':
    s = Solution()
    s.canFinish(2, [[1, 0]])

File: src/main/java/com/baidu/algorithm/dfs/_207_Course_Schedule.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.dfs;

import java.util.*;

/**
 * _207_Course_Schedule
 *
 * @author xuhaoran01
 */
public class _207_Course_Schedule {

    public boolean canFinish(int numCourses, int[][] prerequisites) {

        if (numCourses < 2) {
            return true;
        }

        Map<Integer, List<Integer>> postReq = new HashMap<>();
        for (int[] arr : prerequisites) {
            if (postReq.containsKey(arr[1])) {
                postReq.get(arr[1]).add(arr[0]);
            } else {
                postReq.put(arr[1], new ArrayList<>(Arrays.asList(arr[0])));
            }
        }

        boolean[] visited = new boolean[numCourses];
        for (int i = 0; i < numCourses; i++) {
            if (!visited[i]) {
                if (!dfs(visited, i, postReq)) {
                    return false;
                }
            }
        }

        return true;
    }

    private boolean dfs(boolean[] visited, int idx, Map<Integer, List<Integer>> postReq) {

        visited[idx] = true;
        List<Integer> posts = postReq.get(idx);
        if (posts != null) {
            for (int p : posts) {
                if (visited[p] || !dfs(visited, p, postReq)) {
                    return false;
                }
            }
        }
        visited[idx] = false;

        return true;
    }

    public static void main(String[] args) {
        new _207_Course_Schedule().canFinish(2, new int[][]{{0, 1}});
    }
}
